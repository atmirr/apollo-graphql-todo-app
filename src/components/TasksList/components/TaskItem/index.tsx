import React from 'react';
import { useMutation, useQuery } from '@apollo/client';
import GET_SUB_TASKS from 'src/graphql/queries/getSubTasks';
import UPDATE_TASK from 'src/graphql/mutations/updateTask';
import DELETE_TASK, {
  updateCache as updateCacheAfterDelete,
} from 'src/graphql/mutations/deleteTask';
import MARK_AS_COMPLETED, {
  updateCache as updateCacheAfterCompleted,
} from 'src/graphql/mutations/markAsCompleted';
import MARK_AS_UNCOMPLETED from 'src/graphql/mutations/markAsUncompleted';
import CREATE_TASK, { updateCache } from 'src/graphql/mutations/createTask';
import Task from './components/Task';
import { TaskType } from 'src/types';

function TaskItem({ id, title, completed }: TaskType) {
  const { data: subtasks } = useQuery(GET_SUB_TASKS, {
    variables: { parent_id: id },
  });

  const [editTask, { loading: updateLoading }] = useMutation(UPDATE_TASK);
  const [deleteTask, { loading: deleteLoading }] = useMutation(DELETE_TASK, {
    update(cache, { data: { delete_tasks_by_pk: deletedTask } }) {
      updateCacheAfterDelete(cache, deletedTask);
    },
  });
  const [markAsCompletedTask, { loading: completedLoading }] = useMutation(
    MARK_AS_COMPLETED,
    {
      update(cache, { data: { MarkAsComplete: completedTask } }) {
        updateCacheAfterCompleted(cache, completedTask);
      },
    },
  );
  const [markAsUncompletedTask, { loading: uncompletedLoading }] = useMutation(
    MARK_AS_UNCOMPLETED,
  );
  const [createTask] = useMutation(CREATE_TASK, {
    update(cache, { data: { insert_tasks_one: createdTask } }) {
      updateCache(cache, createdTask);
    },
  });

  const taskIsLoading =
    updateLoading || deleteLoading || completedLoading || uncompletedLoading;

  const handleEdit = ({ id, title }: { id: number; title: string }) =>
    editTask({ variables: { id, title } });
  const handleDelete = (id: number) => deleteTask({ variables: { id } });
  const handleToggleDone = ({
    id,
    completed,
  }: {
    id: number;
    completed: boolean;
  }) =>
    completed
      ? markAsUncompletedTask({ variables: { id } })
      : markAsCompletedTask({ variables: { id } });
  const handleCreateSubtask = ({
    title,
    parentId,
  }: {
    parentId: number;
    title: string;
  }) => createTask({ variables: { title, parent_id: parentId } });
  return (
    <Task
      id={id}
      title={title}
      completed={completed}
      subtasks={subtasks?.readSubTasks}
      handleEdit={handleEdit}
      handleDelete={handleDelete}
      handleToggleDone={handleToggleDone}
      handleCreateSubtask={handleCreateSubtask}
      isLoading={taskIsLoading}
    />
  );
}

export default TaskItem;
