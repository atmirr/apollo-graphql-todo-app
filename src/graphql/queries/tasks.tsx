import { gql } from '@apollo/client';

const TASKS = gql`
  query {
    tasks {
      completed
      id
      parent_id
      title
    }
  }
`;

export default TASKS;
