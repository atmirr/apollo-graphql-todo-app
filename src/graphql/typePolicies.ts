import { Reference, makeVar, TypePolicies } from '@apollo/client';

export const errorVar = makeVar('');

const sortTasks = (
  tasks: Reference[] = [],
  readField: (field: string, ref: Reference) => number,
) =>
  tasks?.sort(
    (a: Reference, b: Reference) =>
      Number(readField('id', b)) - Number(readField('id', a)),
  );

export const typePolicies: TypePolicies = {
  Query: {
    fields: {
      readMainTasks(_, { readField }) {
        const tasks = readField<[]>('tasks');
        const mainTasks = tasks?.filter((taskRef) => {
          const taskParentId = readField('parent_id', taskRef);
          return taskParentId === null;
        });
        return sortTasks(mainTasks, readField);
      },
      readSubTasks(_, { variables, readField }) {
        const tasks = readField<[]>('tasks');
        const subTasks = tasks?.filter((taskRef) => {
          const taskParentId = readField('parent_id', taskRef);
          return taskParentId === variables?.parent_id;
        });
        return sortTasks(subTasks, readField);
      },
      error() {
        return errorVar();
      },
    },
  },
};
