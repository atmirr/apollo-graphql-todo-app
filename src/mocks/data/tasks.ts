const tasks = [
  {
    id: 1,
    title: 'First task',
    completed: false,
    parent_id: null,
    __typename: 'tasks',
  },
  {
    id: 2,
    title: 'Second task',
    completed: true,
    parent_id: null,
    __typename: 'tasks',
  },
  {
    id: 3,
    title: 'A subtask',
    completed: false,
    parent_id: 1,
    __typename: 'tasks',
  },
];

export default tasks;
