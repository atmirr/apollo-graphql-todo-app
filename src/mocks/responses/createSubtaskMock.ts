import CREATE_TASK from 'src/graphql/mutations/createTask';
import tasks from '../data/tasks';
import createdTask from '../data/createdTask';

export const mutationFn = jest.fn();
const createSubtaskMock = {
  request: {
    query: CREATE_TASK,
    variables: { title: createdTask.title, parent_id: tasks[0].id },
  },
  result: () => {
    mutationFn();
    return {
      data: { insert_tasks_one: { ...createdTask, parent_id: tasks[0].id } },
    };
  },
};

export default createSubtaskMock;
