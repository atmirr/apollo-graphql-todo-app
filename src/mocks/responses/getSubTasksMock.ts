import GET_SUB_TASKS from 'src/graphql/queries/getSubTasks';
import tasks from '../data/tasks';
import createdTask from '../data/createdTask';

const getSubTasksMock = {
  request: {
    query: GET_SUB_TASKS,
    variables: { parent_id: tasks[0].id },
  },
  result: {
    data: { readSubTasks: { ...createdTask, parent_id: tasks[0].id } },
  },
};

export default getSubTasksMock;
