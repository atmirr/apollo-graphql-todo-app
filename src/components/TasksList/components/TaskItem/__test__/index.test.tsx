import React from 'react';
import { InMemoryCache } from '@apollo/client';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import {
  waitFor,
  fireEvent,
  queryAllByText,
  queryByLabelText,
} from '@testing-library/react';
import { typePolicies } from 'src/graphql/typePolicies';
import tasks from 'src/mocks/data/tasks';
import updatedTask from 'src/mocks/data/createdTask';
import renderWithMocksGenerator from 'src/mocks/utils/render-with-mocks';
import getSubTasksMock from 'src/mocks/responses/getSubTasksMock';
import createSubtaskMock, {
  mutationFn as createSubtaskMutationFn,
} from 'src/mocks/responses/createSubtaskMock';
import TaskItem from '../index';

const undoneTaskProps = {
  id: tasks[0].id,
  title: tasks[0].title,
  completed: tasks[0].completed,
};

let renderWithMocks;
beforeEach(() => {
  // This way we always have a fresh version
  // of the MockedProvider's cache
  // and our test units wouldn't have the side effect
  const mockedCache = new InMemoryCache({ typePolicies });
  renderWithMocks = renderWithMocksGenerator(mockedCache);
});

describe('<TaskItem />', () => {
  it('should renders the component', () => {
    const { container } = renderWithMocks(<TaskItem {...undoneTaskProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display all the title of the tasks', () => {
    const { container } = renderWithMocks(<TaskItem {...undoneTaskProps} />);
    expect(queryAllByText(container, undoneTaskProps.title).length).toBe(1);
  });

  it('press the add new subtask button should display a input field', () => {
    const { container } = renderWithMocks(<TaskItem {...undoneTaskProps} />);
    const addButton = queryByLabelText(container, 'add');
    fireEvent.click(addButton);
    const inputNode = container.querySelector('input');
    expect(inputNode).toBeVisible();
  });

  it("add a new subtask and press enter should display the task's title", async () => {
    const { container } = renderWithMocks(<TaskItem {...undoneTaskProps} />, [
      createSubtaskMock,
      getSubTasksMock,
    ]);
    const addButton = queryByLabelText(container, 'add');
    fireEvent.click(addButton);
    const inputNode = container.querySelector('input');
    fireEvent.change(inputNode, { target: { value: updatedTask.title } });
    fireEvent.keyDown(inputNode, { keyCode: 13 });
    await waitFor(() => expect(createSubtaskMutationFn).toHaveBeenCalled());
  });

  it("clicking on the edit button should display an input field with the default value of the task's title", () => {
    const { container } = renderWithMocks(<TaskItem {...undoneTaskProps} />);
    const editButton = queryByLabelText(container, 'edit');
    fireEvent.click(editButton);
    const inputNode = container.querySelector('input');
    expect(inputNode.value).toBe(undoneTaskProps.title);
  });
});
