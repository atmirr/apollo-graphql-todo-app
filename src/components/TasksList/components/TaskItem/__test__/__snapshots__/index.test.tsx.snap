// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskItem /> should renders the component 1`] = `
<div
  class="makeStyles-root-2"
  data-testid="task-1"
>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root makeStyles-task-3 MuiListItem-gutters MuiListItem-button"
    role="button"
    tabindex="0"
  >
    <div
      class="makeStyles-wrapper-4"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            aria-label="undone"
            class="MuiSvgIcon-root"
            data-testid="undone-1"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 7.58L10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiListItemText-root makeStyles-listItemText-7"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-8 MuiTypography-body1 MuiTypography-displayBlock"
        >
          First task
        </span>
      </div>
    </div>
    <div
      class="makeStyles-root-10"
    >
      <button
        aria-label="add"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
        data-testid="add-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="edit"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
        data-testid="edit-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="delete"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-edgeEnd"
        data-testid="delete-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
</div>
`;
